### PR TITLE
Interactive Terminal

### PR DESCRIPTION
**Pull request by SmartManoj**
**What is the problem that this fixes or functionality that this introduces? Does it fix any open issues?**

Solves #1655 
image

### CODE REVIEW

**Review by xingyaoww — COMMENTED** [review]
Amazing! This is probably one of the most wanted feature - but i guess it will add some complication on the agent side though.

Just want to confirm:
1. Since the commands are sent directly to the event stream, the agent will be able to see these commands executed by the user, right?
2. Should we consider displaying this differently to the agent (as oppose to agent's own command to avoid confusion)? Something like: "user run command: XXX \n Observation: YYY"?
3. Will the user be able to issue command when the *agent is working*, or they are only allow to run command when the agent is *awaiting user input*? If the former, how would the synchronization look like (e.g., user is typing halfway through a command, and a new command is executed by the agent, hence messed up the terminal display).
4. Can we add some unit/integration tests on the backend side to test for interleave of "user-provided" AND "agent-provided" command?

**Comment by tobitege** [comment]
I'm a bit concerned, that the LLM sees any user input from the terminal, **unless** it is an option to be toggled via the UI.

If this gets inserted into the conversation whilst the LLM is working on a task, this might completely throw it off track of what it was already doing and makes the results unpredictable. Even if the user just does a simple "ls" in terminal.

File: frontend/src/hooks/useTerminal.ts [hunk 1]
```diff
@@ -26,13 +27,54 @@ export const useTerminal = (commands: Command[] = []) => {
...
     if (ref.current) {
       /* Initialize the terminal in the DOM */
       terminal.current.loadAddon(fitAddon.current);
       terminal.current.open(ref.current);
 
       terminal.current.write("$ ");
+      terminal.current.onKey(({ key, domEvent }) => {
+        if (domEvent.key === "Enter") {
+          terminal.current?.write("\r\n");
+          console.log(commandBuffer);
```
Comment (neubig) [inline frontend/src/hooks/useTerminal.ts:41]: Please remove.

File: frontend/src/hooks/useTerminal.ts [hunk 2]
```diff
@@ -26,13 +27,54 @@ export const useTerminal = (commands: Command[] = []) => {
...
+          sendTerminalCommand(commandBuffer);
+          commandBuffer = "";
+        } else if (domEvent.key === "Backspace") {
+          if (commandBuffer.length > 0) {
+            commandBuffer = commandBuffer.slice(0, -1);
+            terminal.current?.write("\b \b");
+          }
+        } else {
+          // console.log(key, key.charCodeAt(0));
```
Comment (neubig) [inline frontend/src/hooks/useTerminal.ts:50]: Please remove.

**Review by neubig — COMMENTED** [review]
This looks great! I agree that there is a bit of a concern that the agent might be confused by the intervening commands, but this seems nice enough as-is that maybe we could worry about that later in a future PR? I just had a few comments about removing console.log functions.

**Review by neubig — APPROVED** [review]

**Pull request merged**